packages/server/src/shared/internal/serialize.ts

```diff
@@ -0,0 +1,62 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+
+/**
+ * @link https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/packages/remix-server-runtime/serialize.ts
+ */
+type JsonPrimitive =
+  | string
+  | number
+  | boolean
+  | String
+  | Number
+  | Boolean
+  | null;
+type NonJsonPrimitive = undefined | Function | symbol;
+
+/*
+ * `any` is the only type that can let you equate `0` with `1`
+ * See https://stackoverflow.com/a/49928360/1490091
+ */
+type IsAny<T> = 0 extends 1 & T ? true : false;
+
+// prettier-ignore
+export type Serialize<T> =
+ IsAny<T> extends true ? any :
+ T extends JsonPrimitive ? T :
+ T extends Map<any,any> | Set<any> ? {} : 
+ T extends NonJsonPrimitive ? never :
+ T extends { toJSON(): infer U } ? U :
+ T extends [] ? [] :
+ T extends [unknown, ...unknown[]] ? SerializeTuple<T> :
+ T extends ReadonlyArray<infer U> ? (U extends NonJsonPrimitive ? null : Serialize<U>)[] :
+ T extends object ? SerializeObject<UndefinedToOptional<T>> :
+ never;
+
+/** JSON serialize [tuples](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) */
+type SerializeTuple<T extends [unknown, ...unknown[]]> = {
+  [k in keyof T]: T[k] extends NonJsonPrimitive ? null : Serialize<T[k]>;
+};
+
+/** JSON serialize objects (not including arrays) and classes */
+type SerializeObject<T extends object> = {
+  [k in keyof T]: T[k] extends NonJsonPrimitive ? never : Serialize<T[k]>;
+};
```

Comment:
right now any keys of type `NonJsonPrimitive` still show up as `never` because when I tried applying the filter keys strategy, TS broke